instrumentation/eclipse-osgi-3.6/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/osgi/EclipseOsgiInstrumentationModule.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.osgi;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static net.bytebuddy.matcher.ElementMatchers.isMethod;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.returns;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.javaagent.instrumentation.api.InClassLoaderMatcher;
+import io.opentelemetry.javaagent.tooling.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import java.util.List;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+/**
+ * The ClassLoaderMatcher's call to ClassLoader.getResource() causes the Eclipse OSGi class loader
+ * to "dynamically import" a bundle for the package if such a bundle is not found, which can lead to
+ * application failure later on due to the bundle hierarchy no longer being "consistent".
+ *
+ * <p>Any side-effect of the ClassLoaderMatcher's call to ClassLoader.getResource() is generally
+ * undesirable, and so this instrumentation patches the behavior and suppresses the "dynamic import"
+ * of the missing package/bundle when the call is originating from ClassLoaderMatcher..
+ */
+@AutoService(InstrumentationModule.class)
+public class EclipseOsgiInstrumentationModule extends InstrumentationModule {
+  public EclipseOsgiInstrumentationModule() {
+    super("eclipse-osgi");
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return singletonList(new EclipseOsgiInstrumentation());
+  }
+
+  private static class EclipseOsgiInstrumentation implements TypeInstrumentation {
+
+    @Override
+    public ElementMatcher<TypeDescription> typeMatcher() {
+      return named("org.eclipse.osgi.internal.loader.BundleLoader");
+    }
+
+    @Override
+    public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+      return singletonMap(
+          isMethod().and(named("isDynamicallyImported")).and(returns(boolean.class)),
+          EclipseOsgiInstrumentation.class.getName() + "$IsDynamicallyImportedAdvice");
+    }
+
+    public static class IsDynamicallyImportedAdvice {
+
+      @Advice.OnMethodEnter(skipOn = Advice.OnNonDefaultValue.class, suppress = Throwable.class)
+      public static boolean onEnter() {
+        return InClassLoaderMatcher.get();
+      }
+
+      @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+      public static void onExit(
+          @Advice.Return(readOnly = false) boolean result,
+          @Advice.Enter boolean inClassLoaderMatcher) {
+        if (inClassLoaderMatcher) {
```

Comment:
Why we cannot call `InClassLoaderMatcher.get()` right here? Why do we need `onEnter`?

Reply:
I think it's because of `skipOn = Advice.OnNonDefaultValue.class`: when `InClassLoaderMatcher.get()` returns `true` the whole method is skipped and in `onExit()` we just make sure that proper value is returned.

Reply:
Wow, this certainly requires documentation :)

Reply:
documentation added 👍